File: server/typo-generator.ts
/**
 * Typo Radius Expansion Module
 * 
 * Generates variations of input phrases to capture common typos:
 * - Keyboard adjacency mistakes (qwerty layout)
 * - Character transpositions (teh → the)
 * - Phonetic substitutions (f→ph, k→c, etc.)
 * - Missing/extra characters
 * - Case variations
 * 
 * Used to expand hypothesis space for recovery when users may have
 * misremembered or mistyped their original passphrase.
 */

/**
 * QWERTY keyboard layout for adjacency detection
 * Maps each key to its neighbors
 */
const KEYBOARD_ADJACENCY: Record<string, string[]> = {
  'q': ['w', 'a', 's'],
  'w': ['q', 'e', 'a', 's', 'd'],
  'e': ['w', 'r', 's', 'd', 'f'],
  'r': ['e', 't', 'd', 'f', 'g'],
  't': ['r', 'y', 'f', 'g', 'h'],
  'y': ['t', 'u', 'g', 'h', 'j'],
  'u': ['y', 'i', 'h', 'j', 'k'],
  'i': ['u', 'o', 'j', 'k', 'l'],
  'o': ['i', 'p', 'k', 'l'],
  'p': ['o', 'l'],
  'a': ['q', 'w', 's', 'z'],
  's': ['a', 'w', 'e', 'd', 'z', 'x'],
  'd': ['s', 'e', 'r', 'f', 'x', 'c'],
  'f': ['d', 'r', 't', 'g', 'c', 'v'],
  'g': ['f', 't', 'y', 'h', 'v', 'b'],
  'h': ['g', 'y', 'u', 'j', 'b', 'n'],
  'j': ['h', 'u', 'i', 'k', 'n', 'm'],
  'k': ['j', 'i', 'o', 'l', 'm'],
  'l': ['k', 'o', 'p'],
  'z': ['a', 's', 'x'],
  'x': ['z', 's', 'd', 'c'],
  'c': ['x', 'd', 'f', 'v'],
  'v': ['c', 'f', 'g', 'b'],
  'b': ['v', 'g', 'h', 'n'],
  'n': ['b', 'h', 'j', 'm'],
  'm': ['n', 'j', 'k'],
  '1': ['2', 'q'],
  '2': ['1', '3', 'q', 'w'],
  '3': ['2', '4', 'w', 'e'],
  '4': ['3', '5', 'e', 'r'],
  '5': ['4', '6', 'r', 't'],
  '6': ['5', '7', 't', 'y'],
  '7': ['6', '8', 'y', 'u'],
  '8': ['7', '9', 'u', 'i'],
  '9': ['8', '0', 'i', 'o'],
  '0': ['9', 'o', 'p'],
};

/**
 * Phonetic substitutions for common sound-alike errors
 */
const PHONETIC_SUBSTITUTIONS: Record<string, string[]> = {
  'ph': ['f'],
  'f': ['ph'],
  'k': ['c', 'ck'],
  'c': ['k', 's'],
  's': ['c', 'z'],
  'z': ['s'],
  'ck': ['k', 'c'],
  'th': ['t'],
  't': ['th'],
  'sh': ['ch'],
  'ch': ['sh'],
  'ee': ['ea', 'ie'],
  'ea': ['ee', 'ia'],
  'ie': ['ee', 'ei'],
  'ei': ['ie'],
  'oo': ['ou', 'u'],
  'ou': ['oo', 'ow'],
  'ow': ['ou'],
  'ai': ['ay', 'a'],
  'ay': ['ai'],
  'tion': ['sion', 'shun'],
  'sion': ['tion', 'shun'],
};

export interface TypoVariation {
  variant: string;
  type: 'keyboard' | 'transposition' | 'phonetic' | 'omission' | 'insertion' | 'case';
  distance: number; // Edit distance from original
  likelihood: number; // 0-1 probability this is the intended phrase
}

/**
 * Generate keyboard adjacency typos
 * Example: "bitcoin" → "bitcoim" (n→m), "bitxoin" (c→x)
 */
export function generateKeyboardTypos(phrase: string, maxDistance: number = 1): TypoVariation[] {
  const variants: TypoVariation[] = [];
  const lowerPhrase = phrase.toLowerCase();
  
  for (let i = 0; i < lowerPhrase.length; i++) {
    const char = lowerPhrase[i];
    if (!char.match(/[a-z0-9]/)) continue;
    
    const neighbors = KEYBOARD_ADJACENCY[char] || [];
    for (const neighbor of neighbors) {
      const variant = lowerPhrase.substring(0, i) + neighbor + lowerPhrase.substring(i + 1);
      variants.push({
        variant,
        type: 'keyboard',
        distance: 1,
        likelihood: 0.3, // Keyboard typos are fairly common
      });
    }
  }
  
  return variants;
}

/**
 * Generate transposition typos (swapped adjacent characters)
 * Example: "satoshi" → "satoish", "bitcoin" → "bitocin"
 */
export function generateTranspositions(phrase: string): TypoVariation[] {
  const variants: TypoVariation[] = [];
  const lowerPhrase = phrase.toLowerCase();
  
  for (let i = 0; i < lowerPhrase.length - 1; i++) {
    const variant = 
      lowerPhrase.substring(0, i) + 
      lowerPhrase[i + 1] + 
      lowerPhrase[i] + 
      lowerPhrase.substring(i + 2);
    
    variants.push({
      variant,
      type: 'transposition',
      distance: 1,
      likelihood: 0.4, // Transpositions are very common
    });
  }
  
  return variants;
}

/**
 * Generate phonetic substitutions
 * Example: "satoshi" → "satoci", "crypto" → "krypto"
 */
export function generatePhoneticVariants(phrase: string): TypoVariation[] {
  const variants: TypoVariation[] = [];
  const lowerPhrase = phrase.toLowerCase();
  
  // Try each phonetic substitution pattern
  for (const [pattern, replacements] of Object.entries(PHONETIC_SUBSTITUTIONS)) {
    let searchPos = 0;
    while (true) {
      const pos = lowerPhrase.indexOf(pattern, searchPos);
      if (pos === -1) break;
      
      for (const replacement of replacements) {
        const variant = 
          lowerPhrase.substring(0, pos) + 
          replacement + 
          lowerPhrase.substring(pos + pattern.length);
        
        variants.push({
          variant,
          type: 'phonetic',
          distance: Math.abs(pattern.length - replacement.length) + 1,
          likelihood: 0.25, // Phonetic errors somewhat common
        });
      }
      
      searchPos = pos + 1;
    }
  }
  
  return variants;
}

/**
 * Generate character omission variants (missing one character)
 * Example: "bitcoin" → "bitcon", "satoshi" → "satshi"
 */
export function generateOmissions(phrase: string, maxOmissions: number = 1): TypoVariation[] {
  const variants: TypoVariation[] = [];
  const lowerPhrase = phrase.toLowerCase();
  
  if (maxOmissions < 1 || lowerPhrase.length < 3) return variants;
  
  for (let i = 0; i < lowerPhrase.length; i++) {
    const variant = lowerPhrase.substring(0, i) + lowerPhrase.substring(i + 1);
    variants.push({
      variant,
      type: 'omission',
      distance: 1,
      likelihood: 0.2, // Missing characters less common but possible
    });
  }
  
  return variants;
}

/**
 * Generate character insertion variants (extra character)
 * Example: "bitcoin" → "bitcooin", "satoshi" → "satooshi"
 */
export function generateInsertions(phrase: string, maxInsertions: number = 1): TypoVariation[] {
  const variants: TypoVariation[] = [];
  const lowerPhrase = phrase.toLowerCase();
  
  if (maxInsertions < 1) return variants;
  
  // Insert doubled characters (most common type)
  for (let i = 0; i < lowerPhrase.length; i++) {
    const char = lowerPhrase[i];
    if (!char.match(/[a-z]/)) continue;
    
    const variant = lowerPhrase.substring(0, i + 1) + char + lowerPhrase.substring(i + 1);
    variants.push({
      variant,
      type: 'insertion',
      distance: 1,
      likelihood: 0.15, // Insertions less common
    });
  }
  
  return variants;
}

/**
 * Generate case variations
 * Example: "Bitcoin" → "bitcoin", "BITCOIN", "BitCoin", "bITCOIN"
 */
export function generateCaseVariations(phrase: string): TypoVariation[] {
  const variants: TypoVariation[] = [];
  
  // All lowercase
  variants.push({
    variant: phrase.toLowerCase(),
    type: 'case',
    distance: 0,
    likelihood: 0.9, // Most likely
  });
  
  // All uppercase
  variants.push({
    variant: phrase.toUpperCase(),
    type: 'case',
    distance: 0,
    likelihood: 0.5,
  });
  
  // Title case (first letter of each word capitalized)
  const titleCase = phrase
    .toLowerCase()
    .split(' ')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
  
  variants.push({
    variant: titleCase,
    type: 'case',
    distance: 0,
    likelihood: 0.6,
  });
  
  // First letter capitalized
  if (phrase.length > 0) {
    variants.push({
      variant: phrase.charAt(0).toUpperCase() + phrase.slice(1).toLowerCase(),
      type: 'case',
      distance: 0,
      likelihood: 0.7,
    });
  }
  
  return variants;
}

/**
 * Generate all typo variations for a phrase
 * Returns sorted by likelihood descending
 */
export function generateAllTypoVariations(
  phrase: string,
  options?: {
    includeKeyboard?: boolean;
    includeTransposition?: boolean;
    includePhonetic?: boolean;
    includeOmission?: boolean;
    includeInsertion?: boolean;
    includeCase?: boolean;
    maxVariants?: number;
  }
): TypoVariation[] {
  const opts = {
    includeKeyboard: true,
    includeTransposition: true,
    includePhonetic: true,
    includeOmission: true,
    includeInsertion: true,
    includeCase: true,
    maxVariants: 100,
    ...options,
  };
  
  const allVariants: TypoVariation[] = [];
  
  if (opts.includeCase) {
    allVariants.push(...generateCaseVariations(phrase));
  }
  
  if (opts.includeTransposition) {
    allVariants.push(...generateTranspositions(phrase));
  }
  
  if (opts.includeKeyboard) {
    allVariants.push(...generateKeyboardTypos(phrase));
  }
  
  if (opts.includePhonetic) {
    allVariants.push(...generatePhoneticVariants(phrase));
  }
  
  if (opts.includeOmission) {
    allVariants.push(...generateOmissions(phrase));
  }
  
  if (opts.includeInsertion) {
    allVariants.push(...generateInsertions(phrase));
  }
  
  // Deduplicate
  const seen = new Set<string>();
  const uniqueVariants = allVariants.filter(v => {
    if (seen.has(v.variant) || v.variant === phrase.toLowerCase()) {
      return false;
    }
    seen.add(v.variant);
    return true;
  });
  
  // Sort by likelihood descending, then by edit distance ascending
  uniqueVariants.sort((a, b) => {
    if (Math.abs(a.likelihood - b.likelihood) > 0.01) {
      return b.likelihood - a.likelihood;
    }
    return a.distance - b.distance;
  });
  
  // Limit to max variants
  return uniqueVariants.slice(0, opts.maxVariants);
}

/**
 * Generate multi-word phrase variations with typos in each word
 * Example: "satoshi nakamoto" → "satoshi nakamato", "satoshi nakamotto", etc.
 */
export function generateMultiWordTypos(phrase: string, maxVariants: number = 50): TypoVariation[] {
  const words = phrase.toLowerCase().split(/\s+/);
  if (words.length === 1) {
    return generateAllTypoVariations(phrase, { maxVariants });
  }
  
  const variants: TypoVariation[] = [];
  
  // Generate typos for each word independently
  for (let i = 0; i < words.length; i++) {
    const wordVariants = generateAllTypoVariations(words[i], { maxVariants: 20 });
    
    for (const variant of wordVariants) {
      const newWords = [...words];
      newWords[i] = variant.variant;
      
      variants.push({
        variant: newWords.join(' '),
        type: variant.type,
        distance: variant.distance,
        likelihood: variant.likelihood * 0.8, // Reduce likelihood slightly for multi-word
      });
    }
  }
  
  // Sort and limit
  variants.sort((a, b) => b.likelihood - a.likelihood);
  return variants.slice(0, maxVariants);
}

/**
 * Levenshtein distance calculation for fuzzy matching
 */
export function levenshteinDistance(str1: string, str2: string): number {
  const len1 = str1.length;
  const len2 = str2.length;
  const matrix: number[][] = [];
  
  for (let i = 0; i <= len1; i++) {
    matrix[i] = [i];
  }
  
  for (let j = 0; j <= len2; j++) {
    matrix[0][j] = j;
  }
  
  for (let i = 1; i <= len1; i++) {
    for (let j = 1; j <= len2; j++) {
      const cost = str1[i - 1] === str2[j - 1] ? 0 : 1;
      matrix[i][j] = Math.min(
        matrix[i - 1][j] + 1,      // deletion
        matrix[i][j - 1] + 1,      // insertion
        matrix[i - 1][j - 1] + cost // substitution
      );
    }
  }
  
  return matrix[len1][len2];
}

File: qig-backend/olympus/typo_generator.py
"""
Typo Generation Module for Python

Generates variations of input phrases to capture common typos:
- Keyboard adjacency mistakes (QWERTY layout)
- Character transpositions (teh → the)
- Phonetic substitutions (f→ph, k→c, etc.)
- Missing/extra characters
- Case variations

Port of server/typo-generator.ts for use in Hephaestus hypothesis generation.
"""

from typing import List, Dict, Tuple
from dataclasses import dataclass


# QWERTY keyboard adjacency map
KEYBOARD_ADJACENCY: Dict[str, List[str]] = {
    'q': ['w', 'a', 's'],
    'w': ['q', 'e', 'a', 's', 'd'],
    'e': ['w', 'r', 's', 'd', 'f'],
    'r': ['e', 't', 'd', 'f', 'g'],
    't': ['r', 'y', 'f', 'g', 'h'],
    'y': ['t', 'u', 'g', 'h', 'j'],
    'u': ['y', 'i', 'h', 'j', 'k'],
    'i': ['u', 'o', 'j', 'k', 'l'],
    'o': ['i', 'p', 'k', 'l'],
    'p': ['o', 'l'],
    'a': ['q', 'w', 's', 'z'],
    's': ['a', 'w', 'e', 'd', 'z', 'x'],
    'd': ['s', 'e', 'r', 'f', 'x', 'c'],
    'f': ['d', 'r', 't', 'g', 'c', 'v'],
    'g': ['f', 't', 'y', 'h', 'v', 'b'],
    'h': ['g', 'y', 'u', 'j', 'b', 'n'],
    'j': ['h', 'u', 'i', 'k', 'n', 'm'],
    'k': ['j', 'i', 'o', 'l', 'm'],
    'l': ['k', 'o', 'p'],
    'z': ['a', 's', 'x'],
    'x': ['z', 's', 'd', 'c'],
    'c': ['x', 'd', 'f', 'v'],
    'v': ['c', 'f', 'g', 'b'],
    'b': ['v', 'g', 'h', 'n'],
    'n': ['b', 'h', 'j', 'm'],
    'm': ['n', 'j', 'k'],
    '1': ['2', 'q'],
    '2': ['1', '3', 'q', 'w'],
    '3': ['2', '4', 'w', 'e'],
    '4': ['3', '5', 'e', 'r'],
    '5': ['4', '6', 'r', 't'],
    '6': ['5', '7', 't', 'y'],
    '7': ['6', '8', 'y', 'u'],
    '8': ['7', '9', 'u', 'i'],
    '9': ['8', '0', 'i', 'o'],
    '0': ['9', 'o', 'p'],
}

# Phonetic substitutions
PHONETIC_SUBSTITUTIONS: Dict[str, List[str]] = {
    'ph': ['f'],
    'f': ['ph'],
    'k': ['c', 'ck'],
    'c': ['k', 's'],
    's': ['c', 'z'],
    'z': ['s'],
    'ck': ['k', 'c'],
    'th': ['t'],
    't': ['th'],
    'sh': ['ch'],
    'ch': ['sh'],
    'ee': ['ea', 'ie'],
    'ea': ['ee', 'ia'],
    'ie': ['ee', 'ei'],
    'ei': ['ie'],
    'oo': ['ou', 'u'],
    'ou': ['oo', 'ow'],
    'ow': ['ou'],
    'ai': ['ay', 'a'],
    'ay': ['ai'],
    'tion': ['sion', 'shun'],
    'sion': ['tion', 'shun'],
}


@dataclass
class TypoVariation:
    """Represents a single typo variation"""
    variant: str
    typo_type: str  # 'keyboard', 'transposition', 'phonetic', 'omission', 'insertion', 'case'
    distance: int  # Edit distance from original
    likelihood: float  # 0-1 probability this is the intended phrase


def generate_keyboard_typos(phrase: str) -> List[TypoVariation]:
    """Generate keyboard adjacency typos"""
    variants = []
    lower_phrase = phrase.lower()
    
    for i, char in enumerate(lower_phrase):
        if not char.isalnum():
            continue
        
        neighbors = KEYBOARD_ADJACENCY.get(char, [])
        for neighbor in neighbors:
            variant = lower_phrase[:i] + neighbor + lower_phrase[i+1:]
            variants.append(TypoVariation(
                variant=variant,
                typo_type='keyboard',
                distance=1,
                likelihood=0.3
            ))
    
    return variants


def generate_transpositions(phrase: str) -> List[TypoVariation]:
    """Generate transposition typos (swapped adjacent characters)"""
    variants = []
    lower_phrase = phrase.lower()
    
    for i in range(len(lower_phrase) - 1):
        variant = (
            lower_phrase[:i] +
            lower_phrase[i+1] +
            lower_phrase[i] +
            lower_phrase[i+2:]
        )
        variants.append(TypoVariation(
            variant=variant,
            typo_type='transposition',
            distance=1,
            likelihood=0.4
        ))
    
    return variants


def generate_phonetic_variants(phrase: str) -> List[TypoVariation]:
    """Generate phonetic substitutions"""
    variants = []
    lower_phrase = phrase.lower()
    
    for pattern, replacements in PHONETIC_SUBSTITUTIONS.items():
        pos = 0
        while True:
            pos = lower_phrase.find(pattern, pos)
            if pos == -1:
                break
            
            for replacement in replacements:
                variant = (
                    lower_phrase[:pos] +
                    replacement +
                    lower_phrase[pos + len(pattern):]
                )
                variants.append(TypoVariation(
                    variant=variant,
                    typo_type='phonetic',
                    distance=abs(len(pattern) - len(replacement)) + 1,
                    likelihood=0.25
                ))
            
            pos += 1
    
    return variants


def generate_omissions(phrase: str) -> List[TypoVariation]:
    """Generate character omission variants (missing one character)"""
    variants = []
    lower_phrase = phrase.lower()
    
    if len(lower_phrase) < 3:
        return variants
    
    for i in range(len(lower_phrase)):
        variant = lower_phrase[:i] + lower_phrase[i+1:]
        variants.append(TypoVariation(
            variant=variant,
            typo_type='omission',
            distance=1,
            likelihood=0.2
        ))
    
    return variants


def generate_insertions(phrase: str) -> List[TypoVariation]:
    """Generate character insertion variants (extra character)"""
    variants = []
    lower_phrase = phrase.lower()
    
    # Insert doubled characters (most common type)
    for i in range(len(lower_phrase)):
        if lower_phrase[i].isalpha():
            variant = lower_phrase[:i+1] + lower_phrase[i] + lower_phrase[i+1:]
            variants.append(TypoVariation(
                variant=variant,
                typo_type='insertion',
                distance=1,
                likelihood=0.15
            ))
    
    return variants


def generate_case_variations(phrase: str) -> List[TypoVariation]:
    """Generate case variations"""
    variants = [
        TypoVariation(variant=phrase.lower(), typo_type='case', distance=0, likelihood=0.9),
        TypoVariation(variant=phrase.upper(), typo_type='case', distance=0, likelihood=0.5),
        TypoVariation(variant=phrase.title(), typo_type='case', distance=0, likelihood=0.6),
    ]
    
    if len(phrase) > 0:
        variants.append(TypoVariation(
            variant=phrase[0].upper() + phrase[1:].lower(),
            typo_type='case',
            distance=0,
            likelihood=0.7
        ))
    
    return variants


def generate_all_typo_variations(
    phrase: str,
    include_keyboard: bool = True,
    include_transposition: bool = True,
    include_phonetic: bool = True,
    include_omission: bool = True,
    include_insertion: bool = True,
    include_case: bool = True,
    max_variants: int = 100
) -> List[TypoVariation]:
    """
    Generate all typo variations for a phrase.
    Returns sorted by likelihood descending.
    """
    all_variants = []
    
    if include_case:
        all_variants.extend(generate_case_variations(phrase))
    
    if include_transposition:
        all_variants.extend(generate_transpositions(phrase))
    
    if include_keyboard:
        all_variants.extend(generate_keyboard_typos(phrase))
    
    if include_phonetic:
        all_variants.extend(generate_phonetic_variants(phrase))
    
    if include_omission:
        all_variants.extend(generate_omissions(phrase))
    
    if include_insertion:
        all_variants.extend(generate_insertions(phrase))
    
    # Deduplicate
    seen = set()
    unique_variants = []
    for variant in all_variants:
        if variant.variant not in seen:
            seen.add(variant.variant)
            unique_variants.append(variant)
    
    # Sort by likelihood descending, then by edit distance ascending
    unique_variants.sort(key=lambda v: (-v.likelihood, v.distance))
    
    # Limit to max variants
    return unique_variants[:max_variants]


def generate_multi_word_typos(phrase: str, max_variants: int = 50) -> List[TypoVariation]:
    """
    Generate multi-word phrase variations with typos in each word.
    Example: "satoshi nakamoto" → "satoshi nakamato", "satoshi nakamotto", etc.
    """
    words = phrase.lower().split()
    if len(words) == 1:
        return generate_all_typo_variations(phrase, max_variants=max_variants)
    
    variants = []
    
    # Generate typos for each word independently
    for i, word in enumerate(words):
        word_variants = generate_all_typo_variations(word, max_variants=20)
        
        for variant in word_variants:
            new_words = words.copy()
            new_words[i] = variant.variant
            
            variants.append(TypoVariation(
                variant=' '.join(new_words),
                typo_type=variant.typo_type,
                distance=variant.distance,
                likelihood=variant.likelihood * 0.8  # Reduce likelihood for multi-word
            ))
    
    # Sort and limit
    variants.sort(key=lambda v: -v.likelihood)
    return variants[:max_variants]


def levenshtein_distance(str1: str, str2: str) -> int:
    """Calculate Levenshtein distance for fuzzy matching"""
    len1, len2 = len(str1), len(str2)
    
    # Create matrix
    matrix = [[0] * (len2 + 1) for _ in range(len1 + 1)]
    
    for i in range(len1 + 1):
        matrix[i][0] = i
    for j in range(len2 + 1):
        matrix[0][j] = j
    
    for i in range(1, len1 + 1):
        for j in range(1, len2 + 1):
            cost = 0 if str1[i-1] == str2[j-1] else 1
            matrix[i][j] = min(
                matrix[i-1][j] + 1,      # deletion
                matrix[i][j-1] + 1,      # insertion
                matrix[i-1][j-1] + cost  # substitution
            )
    
    return matrix[len1][len2]
